feat: use `cached_property` without locks on all supported Python versions

sagarvora opened the pull request:
Closes #32076 

`no-docs`

sagarvora committed:
feat: use `cached_property` without locks on all supported Python versions

ankush commented on frappe/utils/__init__.py:
```diff
@@ -664,9 +664,51 @@ def is_markdown(text):
 		return not NON_MD_HTML_PATTERN.search(text)
 
 
+class cached_prop:
```
Should we use the default for 3.12+? :eyes:

sagarvora replied:
Ok I will implement it.

sagarvora commented on frappe/utils/__init__.py:
```diff
@@ -664,9 +664,57 @@ def is_markdown(text):
 		return not NON_MD_HTML_PATTERN.search(text)
 
 
+class cached_property:
```
I'm keeping the name same so that we can switch without much hassle when we stop supporting 3.11 or lower.

sagarvora replied:
Also inheriting from functools.cached_property to support `isinstance(x, functools.cached_property)` calls outside framework and `__class_getitem__`

ankush committed:
test: update constant for bundle and memory usage

Over time code size has grown without any major regressions, so just
accomodating them here.
